stan/math/gpu/matrix_gpu.hpp

```diff
@@ -0,0 +1,283 @@
+    if (size() > 0) {
+      cl::Context& ctx = opencl_context.context();
+      cl::CommandQueue& queue = opencl_context.queue();
+      try {
+        // creates the OpenCL buffer to copy the Eigen
+        // matrix to the OpenCL device
+        oclBuffer_
+            = cl::Buffer(ctx, CL_MEM_READ_WRITE, sizeof(double) * A.size());
+        /** Writes the contents of A to the OpenCL buffer
+         * starting at the offset 0.
+         * CL_TRUE denotes that the call is blocking as
+         * we do not want to execute any further kernels
+         * on the device until we are sure that the data
+         * is finished transfering)
+         */
+        queue.enqueueWriteBuffer(oclBuffer_, CL_TRUE, 0, sizeof(T) * A.size(),
+                                 A.data());
+      } catch (const cl::Error& e) {
+        check_opencl_error("matrix constructor", e);
+      }
+    }
+  }
+
+  matrix_gpu& operator=(const matrix_gpu& a) {
+    check_size_match("assignment of GPU matrices", "source.rows()", a.rows(),
+                     "destination.rows()", rows());
+    check_size_match("assignment of GPU matrices", "source.cols()", a.cols(),
+                     "destination.cols()", cols());
+    oclBuffer_ = a.buffer();
+    return *this;
+  }
+};
+
+/**
+ * Copies the source Eigen matrix to
+ * the destination matrix that is stored
+ * on the GPU.
+ *
+ * @tparam T type of data in the Eigen matrix
+ * @param src source Eigen matrix
+ * @param dst destination matrix on the GPU
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ *
+ */
+template <typename T, int R, int C>
+void copy(const Eigen::Matrix<T, R, C>& src, matrix_gpu& dst) {
+  check_size_match("copy (Eigen -> GPU)", "src.rows()", src.rows(),
+                   "dst.rows()", dst.rows());
+  check_size_match("copy (Eigen -> GPU)", "src.cols()", src.cols(),
+                   "dst.cols()", dst.cols());
+  if (src.size() > 0) {
+    cl::CommandQueue queue = opencl_context.queue();
+    try {
+      /** writes the contents of src to the OpenCL buffer
+       * starting at the offset 0
+       * CL_TRUE denotes that the call is blocking
+       * We do not want to execute any further kernels
+       * on the device until we are sure that the data is transferred)
+       */
+      queue.enqueueWriteBuffer(dst.buffer(), CL_TRUE, 0, sizeof(T) * dst.size(),
+                               src.data());
+    } catch (const cl::Error& e) {
+      check_opencl_error("copy Eigen->GPU", e);
+    }
+  }
+}
+
+/**
+ * Copies the source matrix that is stored
+ * on the GPU to the destination Eigen
+ * matrix.
+ *
+ * @tparam T type of data in the Eigen matrix
+ * @param src source matrix on the GPU
+ * @param dst destination Eigen matrix
+ *
+ * @throw <code>std::invalid_argument</code> if the
+ * matrices do not have matching dimensions
+ *
+ */
+template <typename T, int R, int C>
+void copy(matrix_gpu& src, Eigen::Matrix<T, R, C>& dst) {
```

Comment:
These `copy` functions should all go into its own header in `stan/math/prim/mat/copy.hpp`.